dali/pipeline/workspace/workspace_data_factory.h

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_WORKSPACE_WORKSPACE_DATA_FACTORY_H_
+#define DALI_PIPELINE_WORKSPACE_WORKSPACE_DATA_FACTORY_H_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "dali/pipeline/graph/op_graph.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/workspace/device_workspace.h"
+#include "dali/pipeline/workspace/host_workspace.h"
+#include "dali/pipeline/workspace/mixed_workspace.h"
+#include "dali/pipeline/workspace/support_workspace.h"
+
+#include "dali/kernels/static_switch.h"
+#include "dali/kernels/tuple_helpers.h"
+#include "dali/pipeline/util/op_type_to_workspace.h"
+
+namespace dali {
+
+
+template <typename Backend>
+void SetPinned(SupportWorkspace::output_t<Backend> &t, bool pinned) {
+  t->set_pinned(pinned);
+}
+
+template <typename Backend>
+void SetPinned(HostWorkspace::output_t<Backend> &t, bool pinned) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->set_pinned(pinned);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+void SetPinned(DeviceWorkspace::output_t<Backend> &t, bool pinned) {
+  t->set_pinned(pinned);
+}
+
+template <typename Backend>
+bool IsPinned(SupportWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+bool IsPinned(HostWorkspace::output_t<Backend> &t) {
+  bool is_pinned = true;
+  for (auto &tensor_ptr : t) {
+    is_pinned = is_pinned && tensor_ptr->is_pinned();
+  }
+  return is_pinned;
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+bool IsPinned(DeviceWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
```

Comment:
why do you need the batch_size?

Reply:
The batch size is needed when we work with CPU Op outputs. The are vector of `batch_size` Tensors.